### PR TITLE
Solution for Conway's Game of Life

### PR DESCRIPTION
**Pull request by fgs-pwright**
This solution includes a game that will be initialized with a square-tile grid and the standard Game of Life rules ("B2/S23").  Different rules could be set for the game, if desired.

Different implementations of the `IGrid` interface could be added to allow the game to be played on different types of surfaces, like a flat hexagonal grid, or even a polyhedron like a soccer ball.

### CODE REVIEW

**Comment by fgs-pwright** [comment]
I'd like to generalize the game a little more by making the use of the `LifeState` enum generic in a few places, but I ran into some snags when I tried.  I may try to get some help with that.

File: GameOfLife/GameOfLife/SquareTileGrid.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        {
+            var rowsInNeighborhood = _rowDimension.GetNeighborValues(centerCellPosition.DimensionOne);
+            var columnsInNeighborhood = _columnDimension.GetNeighborValues(centerCellPosition.DimensionTwo);
+            foreach (var position in GetPositionsExcluding(rowsInNeighborhood, columnsInNeighborhood, centerCellPosition))
+                yield return position;
+
+            if (_rowDimension.IsOwnNeighbor())
+            {
+                yield return centerCellPosition;
+                yield return centerCellPosition;
```
Comment (FGS-BMarkham) [inline GameOfLife/GameOfLife/SquareTileGrid.cs]: I would expect `GetNeighborPositions` to return a _distinct_ set of `CellPosition`, for which this would surprise me. For cases of `IsOwnNeighbor`, I would suggest yielding the `centerCellPosition` a maximum of 1 times.

File: GameOfLife/GameOfLife/SquareTileGrid.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            }
+
+            if (_columnDimension.IsOwnNeighbor())
+            {
+                yield return centerCellPosition;
+                yield return centerCellPosition;
+            }
+        }
+
+        private IEnumerable<CellPosition> GetPositionsExcluding(
```
Comment (FGS-BMarkham) [inline GameOfLife/GameOfLife/SquareTileGrid.cs]: Challenge:
LINQ has an `Exclude<T>(…)` method, which relies on equality. If you implement equality for `CellPosition`, you could probably simplify this down to a string of library calls that doesn't involve you doing your own loops.

File: GameOfLife/GameOfLife/ConsoleWriter.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+
+namespace GameOfLife
+{
+    public class ConsoleWriter : IConsoleWriter
```
Comment (FGS-BMarkham) [inline GameOfLife/GameOfLife/ConsoleWriter.cs]: Good eye, introducing a seam here. Please note that this implementation doesn't have to use the `Console` directly. `System.Console.Out` and `System.Console.In` both return text stream manipulation abstractions that are super important in .NET, and you could just as well make your `IConsoleWriter` implementation take both of those.

Also, if you are going to have a `ReadLine`, better rename that interface to either `IConsole` or `IConsoleReaderWriter`.

File: GameOfLife/GameOfLife/Game.cs [hunk 4]
```diff
@@ -1,6 +1,116 @@
...
     {
+        private IRuleFactory _ruleFactory;
+        private IGridFactory _gridFactory;
+        private IConsoleWriter _consoleWriter;
+        private IRuleset _rules;
+        private IGrid _grid;
+        private List<LifeState[,]> _generations;
+
+        public const string GRID_NOT_LOADED = "A grid must be loaded before any steps can be taken.";
+        public const string INCOMPLETE_RULES = "Cannot step until rules have been defined for all possible cell states.";
```
Comment (FGS-BMarkham) [inline GameOfLife/GameOfLife/Game.cs]: Why are these public?

File: GameOfLife/GameOfLife/Game.cs [hunk 5]
```diff
@@ -1,6 +1,116 @@
-﻿namespace GameOfLife
+﻿using System.Collections.Generic;
+
+namespace GameOfLife
 {
-    public class Game
+    public class Game : IGame
```
Comment (FGS-BMarkham) [inline GameOfLife/GameOfLife/Game.cs]: This class has too many responsibilities. It's interacting with the user, it's managing engine initialization & execution. Please try to find a way to tear out these responsibilities so this class can be higher-level.

File: GameOfLife/GameOfLife/SquareTileGrid.cs [hunk 6]
```diff
@@ -15,13 +16,29 @@ public SquareTileGrid(Cell[,] cells, bool wrapsOnRows, bool wrapsOnColumns)
...
+            foreach (var state in pattern)
+            {
+                consoleWriter.Write(DefaultSettings.ToString(state));
+                column++;
+                if (column > _columnDimension.Max)
+                {
+                    consoleWriter.Write(Environment.NewLine);
+                    column = 0;
+                }
+            }
```
Comment (FGS-BMarkham) [inline GameOfLife/GameOfLife/SquareTileGrid.cs]: This does not fit the responsibility paradigm of the class it was added to. Please consider implementing this elsewhere.

File: GameOfLife/GameOfLife/IGrid.cs [hunk 7]
```diff
@@ -0,0 +1,11 @@
+﻿using System.Collections.Generic;
+
+namespace GameOfLife
+{
+    public interface IGrid
+    {
+        Cell GetCellAt(CellPosition cellPosition);
+        IEnumerable<Cell> GetNeighborsOfCellAt(CellPosition centerCellPosition);
+        IEnumerator<CellPosition> GetEnumerator();
```
Comment (fgs-pwright) [inline GameOfLife/GameOfLife/IGrid.cs]: Thanks.  In order to make this work, will I need to create a new class that implements `IEnumerator`, like the example given [here](https://msdn.microsoft.com/en-us/library/system.collections.ienumerable.getenumerator(v=vs.110).aspx#Anchor_2)?

Comment (FGS-BMarkham) [inline GameOfLife/GameOfLife/IGrid.cs]: Good question! To answer it: Nope. I'm pretty sure that if you use the `yield` syntax, your method signature can either return `IEnumerable<T>` or `IEnumerator<T>`, and the compiler does the magic for you.

File: GameOfLife/GameOfLife/GridFactory.cs [hunk 8]
```diff
@@ -1,18 +1,30 @@
-﻿using System;
-using static System.Math;
-
-namespace GameOfLife
+﻿namespace GameOfLife
 {
     public class GridFactory
     {
-        public Grid Create(LifeState[,] seed, params bool[] inputWrappingRules)
+        public SquareTileGrid CreateSquareTileGrid(LifeState[,] seed)
```
Comment (fgs-pwright) [inline GameOfLife/GameOfLife/GridFactory.cs]: This is true.  My thought is that this would be a simple, working case that didn't require some of the complex validation you mentioned, and that the class could be extended with new `Create` methods, if that would be useful.

I'm going back and forth between thinking this is laziness on my part, or just thinking that this is a straightforward implementation that I can build on in the future.  What do you think?

File: GameOfLife/GameOfLife/Rule.cs [hunk 9]
```diff
@@ -1,4 +1,6 @@
-﻿using System.Linq;
+﻿using System.Collections.Generic;
+using System.Linq;
+using static GameOfLife.LifeState;
```
Comment (fgs-pwright) [inline GameOfLife/GameOfLife/Rule.cs]: The `using static ...` syntax is new?  I've mostly only used it for my enums so far, and I don't think I've been completely consistent with it.  I'll keep in mind that it's new as I'm using it in the future.

File: GameOfLife/GameOfLife/IGrid.cs (hunk 7 above)
Comment (fgs-pwright) [inline GameOfLife/GameOfLife/IGrid.cs]: Thanks again for this point.  It took me some forum scouring to put it together on my first try, and I'm not surprised I didn't do it exactly the best way.  How does it look now?

File: GameOfLife/GameOfLife/Rule.cs (hunk 9 above)
Comment (FGS-BMarkham) [inline GameOfLife/GameOfLife/Rule.cs]: For the first 14 years or so of C#, if you wanted to use another type's static members (enums too), you had to spell out the type names by hand. Each time. It sounds tedious, but some of us are used to _reading_ code like that. Highlights weird dependencies well. Like I said, I'll get over it. :-)

File: GameOfLife/GameOfLife/GridFactory.cs (hunk 8 above)
Comment (FGS-BMarkham) [inline GameOfLife/GameOfLife/GridFactory.cs]: In the land of "user picks which kind of simulation to run", and options include:
- Non-cartesian systems
- Infinite planes (not representable by an array type)
- Variations in dimensions
… this code won't work. That said, none of those are in your requirements right now. Like I mentioned the other day - this is something you might want to come back to during CoP time.

Comment (fgs-pwright) [inline GameOfLife/GameOfLife/GridFactory.cs]: Good cases to consider.  I will almost certainly do that.

File: GameOfLife/GameOfLife/Rule.cs (hunk 9 above)
Comment (fgs-pwright) [inline GameOfLife/GameOfLife/Rule.cs]: That makes sense.  If you see me abusing this to obscure unhealthy dependencies, then let me know.  I'll keep that it mind when I use something like this in the future.

File: GameOfLife/GameOfLife/LifeState.cs [hunk 10]
```diff
@@ -2,7 +2,7 @@
 {
     public enum LifeState
     {
-        Alive,
-        Dead
+        Dead,
+        Alive
```
Comment (fgs-pwright) [inline GameOfLife/GameOfLife/LifeState.cs]: I did this so that `Dead` would be the default value, for a number of reasons.  I'm happy to share some of them, if that would be a fruitful discussion.

File: GameOfLife/GameOfLife/SquareTileGrid.cs (hunk 6 above)
Comment (fgs-pwright) [inline GameOfLife/GameOfLife/SquareTileGrid.cs]: I didn't have this method in this class initially, but I liked the idea of this class being the single source of information about the arrangement of cells or states, which is why I stuck it in here.

I moved this logic to the `Write...` method of the `Game` class, but I was thinking of creating a separate class altogether that would handle this.  Do you think that would be a better way to deal with it?

File: GameOfLife/GameOfLife/Game.cs (hunk 5 above)
Comment (fgs-pwright) [inline GameOfLife/GameOfLife/Game.cs]: Oh, I missed this comment the first time through.  I can see what you're saying, and I'll revise my solution to correct this.

File: GameOfLife/GameOfLife/LifeState.cs (hunk 10 above)
Comment (FGS-BMarkham) [inline GameOfLife/GameOfLife/LifeState.cs]: If you have engine rules or initialization behavior that indirectly relies on one of these values in a specific way, it is best to capture that in a named variable close to where it is being used. For example, mayhaps `GridFactory` should have a `DefaultValue` member `const` that points to `LifeState.Dead`. This will keep your code resilient against changes to that enum in the future.

File: GameOfLife/GameOfLife/SquareTileGrid.cs (hunk 6 above)
Comment (FGS-BMarkham) [inline GameOfLife/GameOfLife/SquareTileGrid.cs]: Separate class is best, methinks.

File: GameOfLife/GameOfLife/IConsole.cs [hunk 11]
```diff
@@ -1,7 +1,11 @@
-﻿namespace GameOfLife
+﻿using System.IO;
+
+namespace GameOfLife
 {
-    public interface IConsoleWriter
+    public interface IConsole
     {
+        TextReader In { get; set; }
+        TextWriter Out { get; set; }
```
Comment (FGS-BMarkham) [inline GameOfLife/GameOfLife/IConsole.cs]: You wouldn't put this on your interface, because it is an implementation detail, ain't it?

File: GameOfLife/GameOfLife/IGame.cs [hunk 12]
```diff
@@ -9,6 +9,6 @@ public interface IGame
         void SetRuleFor(LifeState state, params uint[] neighborCountsYieldingLive);
         void Step();
         void WriteCurrentPatternToConsole();
-        IEnumerable<LifeState> GetCurrentPattern();
+        IEnumerable<IEnumerable<LifeState>> GetCurrentPattern();
```
Comment (FGS-BMarkham) [inline GameOfLife/GameOfLife/IGame.cs]: When I saw this, it reminded me of a series of blog posts I wanted to share with the team: https://foxguard.slack.com/archives/dev-learning-emporium/p1485984405000172

Unless you are doing something really weird here (on purpose), you probably don't need an `IEnumerable<IEnumerable<T>>`. If you did this because you felt cornered into it by the structure of the implementation underneath, you should know that you can use the LINQ method `SelectMany` to flatten it down, or "unwrap" the monad, so-to-speak.

File: GameOfLife/GameOfLife/Game.cs [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    {
+        private const string GRID_NOT_LOADED = "A grid must be loaded before any steps can be taken.";
+        private const string INCOMPLETE_RULES = "Cannot step until rules have been defined for all possible cell states.";
+
+        private IRuleFactory _ruleFactory;
+        private IGridFactory _gridFactory;
+        private IConsole _console;
+        private IRuleset _rules;
+        private IGrid _grid;
+        private List<IEnumerable<IEnumerable<LifeState>>> _generations;
```
Comment (FGS-BMarkham) [inline GameOfLife/GameOfLife/Game.cs]: I'm not sure what this is supposed to be, but it kinda looks like you are saving a complete history of the run of the simulation. This won't work out well under continuous (or long-term) execution, and should probably be avoided.

The `IEnumerable<IEnumerable<LifeState>>` inside the list looks like an attempt to represent the state of a single generation of the grid without revealing the concrete types. Once again, this kinda reveals the jagged array implementation, and seems to be begging for a different kind of abstraction. (For example, what if you had an `IGameState` interface, which represented an immutable snapshot of gamestate? It would probably borrow elements from elsewhere in your application, so it would require some shuffling.)

File: GameOfLife/GameOfLife/Game.cs [hunk 14]
```diff
@@ -0,0 +1,135 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+
+namespace GameOfLife
+{
+    public class Game : IGame
```
Comment (FGS-BMarkham) [inline GameOfLife/GameOfLife/Game.cs]: Initializing, mutating, & validating a set of rules could probably exist in a class with a name more specific to that purpose. The word "Game" here might as well be as vague as "Program", so we need to justify its existence at every turn, and pull things out where we can.

File: GameOfLife/GameOfLifeConsole/Program.cs [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace GameOfLifeConsole
+{
+    public class Program
+    {
+        static void Main(string[] args)
+        {
+            var consoleWriter = new TextReaderWriter();
+            var game = new Game(new RuleFactory(), new GridFactory(), consoleWriter, new Ruleset());
+            var consoleController = new ConsoleInterface(consoleWriter, game, new TextCommandParser(), new CommandRunner());
```
Comment (FGS-BMarkham) [inline GameOfLife/GameOfLifeConsole/Program.cs]: If only there existed some "thing" that could wire all these dependencies together for you, and inject them when you ask for them…
:-)
That can wait until Monopoly, but you'll want to do it early.

File: GameOfLife/GameOfLifeTests/ConsoleInterfaceTests.cs [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+            _consoleMock.Setup(c => c.ReadLine()).Returns(Fixture.Create<string>());
+            _commandParserMock.Setup(p => p.ParseCommand(It.IsAny<string>())).Returns(Command.Quit);
+            _consoleInterface.Start();
+
+            VerifyDisplaysWelcomeMessage();
+            VerifyLoadsNewGame();
+            VerifyDisplaysGrid();
+            VerifyPromptsUserForCommand();
+            VerifyParsesUserCommand();
+        }
```
Comment (FGS-BMarkham) [inline GameOfLife/GameOfLifeTests/ConsoleInterfaceTests.cs]: Great use of decoupling and mocking to make sure your program behaves as expected!

File: GameOfLife/GameOfLifeTests/GameTests.cs [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+        [SetUp]
+        public void SetUp()
+        {
+            var ruleFactory = SetUpMockIRuleFactory();
+            var gridFactory = SetUpMockIGridFactory();
+            var textWriter = SetUpMockTextWriter();
+            var console = SetUpMockIConsole(textWriter);
+            var ruleset = SetUpMockIRuleset();
+
+            Game = new Game(ruleFactory, gridFactory, console, ruleset);
```
Comment (FGS-BMarkham) [inline GameOfLife/GameOfLifeTests/GameTests.cs]: AutoFixture can create your test subjects, too. (Not just random values.) This is a really powerful thing that we use a lot in our projects.

We do this by leaning on two mechanisms.

The first mechanism we lean on is the fact that AutoFixture lets you `.Register(…)` things. When you do this, it will use that thing you registered anytime it encounters a constructor/property/etc that asks for that thing. This is really valuable if you want to wire up a mock that you've just spent pains configuring.

The second mechanism we lean on is the `AutoFixture.AutoMoq` NuGet package, (wire it up with a `Customize` call like you can see [here](http://blog.ploeh.dk/2010/08/19/AutoFixtureasanauto-mockingcontainer/)). This tells AutoFixture that every time it sees an interface or an abstract type in the constructor of a thing, if it doesn't know how to make one, to generate a mock for us.

This makes it really easy for our tests to mock what we need, ignore everything else, and not couple our tests to the layout of the constructor of the type under test.

Give it a poke!

File: GameOfLife/GameOfLifeConsole/Program.cs (hunk 15 above)
Comment (fgs-pwright) [inline GameOfLife/GameOfLifeConsole/Program.cs]: Ha, yep.  Is there a DI framework that the team uses commonly, or one you'd suggest I check out?

Comment (fgs-dbudwin) [inline GameOfLife/GameOfLifeConsole/Program.cs]: We tend to use Autofac.  We have used Ninject in the past though.

Comment (fgs-pwright) [inline GameOfLife/GameOfLifeConsole/Program.cs]: Cool, thanks

Comment (fgs-dbudwin) [inline GameOfLife/GameOfLifeConsole/Program.cs]: DI is weird at first...then you can't live without it.  Autofac is a bit of a more mature product than Ninject, but Ninject is slightly easier to work with.  You'll notice they have a lot of analogous actions between them.  I did my kata with Ninject but wished I had done it Autofac since that's what we actually use.  I would say it's only slightly more difficult to learn, especially for straight forward applications like katas.  YMMV.  We do still _technically_ use Ninject for 1 or 2 of our ~50 projects in PUMP.  But those projects are rarely touched which is why they haven't been switched over yet.  Let us know if you have any questions about this stuff.

File: GameOfLife/GameOfLifeConsole/Program.cs [hunk 18]
```diff
@@ -1,20 +1,36 @@
...
         static void Main(string[] args)
         {
-            var consoleReaderWriter = new ConsoleReaderWriter();
-            var commandRunner = new CommandRunner(
-                new Ruleset(new RuleFactory()),
-                new GridFactory(),
-                new GameAdvancer(consoleReaderWriter),
-                new GridWriter(consoleReaderWriter));
+            var builder = new ContainerBuilder();
+            builder.RegisterType<ConsoleReaderWriter>().As<IConsoleReaderWriter>().As<IConsoleWriter>();
```
Comment (fgs-dbudwin) [inline GameOfLife/GameOfLifeConsole/Program.cs]: This looks pretty on point with how we setup Autofac!

However, remember we like short `Main` methods that don't do much.  One suggestion is to take all of your registration out into a new file.  Our naming convention is `[Something]Module.cs`.  We often have many modules per project, but I think you can get away with one for simplicity sake.  Make the class inherit `Module` which is an Autofac class.  You'll be forced to re-implement a `Load` function where you can put all of your registrations.

File: GameOfLife/GameOfLifeConsole/Program.cs [hunk 19]
```diff
@@ -1,20 +1,36 @@
...
+            builder.RegisterType<RuleFactory>().As<IRuleFactory>();
+            builder.RegisterType<Ruleset>().As<IRuleset>();
+            builder.RegisterType<GridFactory>().As<IGridFactory>();
+            builder.RegisterType<GameAdvancer>().As<IGameAdvancer>();
+            builder.RegisterType<SquareTileGridWriter>().As<IGridWriter>();
+            builder.RegisterType<CommandRunner>().As<ICommandRunner>();
+            builder.RegisterType<TextCommandParser>().As<TextCommandParser>();
```
Comment (fgs-dbudwin) [inline GameOfLife/GameOfLifeConsole/Program.cs]: When you register a type against itself, you can use `AsSelf()`: ` builder.RegisterType<TextCommandParser>().AsSelf()`

File: GameOfLife/GameOfLifeConsole/Program.cs [hunk 20]
```diff
@@ -1,20 +1,36 @@
...
+        private static IContainer Container { get; set; }
+
         static void Main(string[] args)
         {
-            var consoleReaderWriter = new ConsoleReaderWriter();
-            var commandRunner = new CommandRunner(
-                new Ruleset(new RuleFactory()),
-                new GridFactory(),
-                new GameAdvancer(consoleReaderWriter),
-                new GridWriter(consoleReaderWriter));
```
Comment (fgs-dbudwin) [inline GameOfLife/GameOfLifeConsole/Program.cs]: Where did all this injecting disappear to?  I don't see it in this commit...did it get moved into another commit?

Comment (fgs-dbudwin) [inline GameOfLife/GameOfLifeConsole/Program.cs]: ☝️ Ignore, haz dumbz

File: GameOfLife/GameOfLifeConsole/Program.cs (hunk 18 above)
Comment (fgs-pwright) [inline GameOfLife/GameOfLifeConsole/Program.cs]: Good deal.  What's the significance of the "Something" in `[Something]Module.cs`?  I'm wondering what I should consider when I name this one.  Maybe something that captures the most high-level class registered, like `ConsoleGameModule.cs`?

**Comment by FGS-BMarkham** [comment]
Once upon a time, I gave feedback on a kata to the effect of: the Main method should be 4ish lines of code. Allow me to demonstrate it again:
```
using (var container = MySuperSpecialDependencyInjectionContainerFactory.Create())
{
    var mySpecialStartupObject = container.Resolve<IMySpecialStartupType>();
    mySpecialStartupObject.DoSomething();
}
```

Shoot for this.

**Comment by fgs-pwright** [comment]
Ok, I moved my registration logic to a testable module, added a test class for it, and registered that module in a `Create` method of a static container factory class.

I wondered if I could test module registration separate from service resolution (ie. test that the factory `Create` method yielded a container with the module registered, and separately test that a container with that module registered resolved services as defined in the module), but it didn't look like a container would store any information about the _modules_ registered in it, but only the components.

> [Modules (...) are used to configure the container, they are not actually registered (...) like other components](http://autofac.readthedocs.io/en/latest/configuration/modules.html)

If there is a way I can or should test these separately, I'd be glad to do it but may need some hints.

**Comment by FGS-BMarkham** [comment]
I have no further review comments for this PR.

However, this PR is unmergeable due to the target branch being `master`, and that not being the convention for our CodeCamp PRs. That makes :gopher: :squash: moot.

We'll just close the PR and pretend like we merged it. :-)
